### PR TITLE
ENH: implement matching warning message

### LINKED ISSUE #37261: ENH: match warning messages
#### Is your feature request related to a problem?

I wish we could match warning messages using ``pandas._testing.assert_produces_warning``.
Currently it catches the warning category, but does not match the message.
Recently (https://github.com/pandas-dev/pandas/pull/36982) I stumbled upon the need for this feature.

#### Describe the solution you'd like

``pandas._testing.assert_produces_warning`` should take optional keyword argument ``match``, to make it look similar to ``pytest.warns``.

#### API breaking implications

Should not break existing API, as ``match`` would be an optional argument.

#### Describe alternatives you've considered

I used custom way of catching warnings (via ``warnings.catch_warnings``) and matching their messages, but was suggested to give preference to ``assert_produces_warning``.

### PR DESCRIPTION
**Pull request by ivanovmg**
- [ ] closes #37261
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Implement matching warning message.
New kwarg ``match`` in ``assert_produces_warning``.

### CODE REVIEW

File: pandas/_testing.py [hunk 1]
```diff
@@ -2550,6 +2551,7 @@ def assert_produces_warning(
     filter_level="always",
     check_stacklevel=True,
     raise_on_extra_warnings=True,
+    match=None,
```
Comment (jbrockmendel) [inline pandas/_testing.py:2554]: can you annoate

File: pandas/tests/_testing/test_warnings.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        ResourceWarning,
+        UserWarning,
+        FutureWarning,
+        DeprecationWarning,
+        PerformanceWarning,
+        DtypeWarning,
+    ],
+)
+def category(request):
+    """Return unique warning.
```
Comment (jbrockmendel) [inline pandas/tests/_testing/test_warnings.py:26]: newline after """

File: pandas/_testing.py [hunk 3]
```diff
@@ -2641,18 +2643,37 @@ class for all warnings. To check that no warning is returned,
...
+                f"matching {match}"
+            )
+
         if raise_on_extra_warnings and extra_warnings:
             raise AssertionError(
                 f"Caused unexpected warning(s): {repr(extra_warnings)}"
             )
 
 
+def _assert_raised_with_correct_stacklevel(actual_warning):
```
Comment (jreback) [inline pandas/_testing.py:2665]: bonus points for typing

**Review by jreback — COMMENTED** [review]
comment, but ok to merge after @jbrockmendel doc-string comments.

File: pandas/_testing.py (hunk 1 above)
Comment (ivanovmg) [reply pandas/_testing.py]: Annotated, but presumably the code became slightly less pythonic. Sorry for that.

File: pandas/_testing.py (hunk 3 above)
Comment (ivanovmg) [reply pandas/_testing.py]: Annotated. IMHO, the type ``warnings.WarningMessage``, although correct, has a confusing name for warning instance.

File: pandas/tests/_testing/test_warnings.py (hunk 2 above)
Comment (ivanovmg) [reply pandas/tests/_testing/test_warnings.py]: Fixed.

File: pandas/tests/_testing/test_warnings.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
+""""
+Test module for testing ``pandas._testing.assert_produces_warning``.
+"""
```
Comment (jorisvandenbossche) [inline pandas/tests/_testing/test_warnings.py:3]: There is already a file with tests for this, in `tests/util/test_assert_produces_warning.py`, can you move this there?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for working on this!

File: pandas/tests/_testing/test_warnings.py (hunk 4 above)
Comment (ivanovmg) [reply pandas/tests/_testing/test_warnings.py]: Oh, I completely missed it. I moved tests as requested.

**Comment by jreback** [comment]
thanks @ivanovmg very nice

**Pull request merged**